### ISSUE TITLE
Potential rework of autodiff API

### ISSUE THREAD

**Issue by nkoep** [pymanopt/pymanopt#55]
While looking into the changes necessary to add a pytorch autodiff backend (cf. #52), I've noticed some deficiencies with our current autodiff implementation. The most important one that currently prevents us from simply adding a pytorch backend module is the fact that both pytorch and autograd build their computational graphs dynamically. As such, both backends expect the cost function to be a regular callable, and determine gradients and higher-order derivatives for a given input during execution.

In our current implementation, we always assume a callable cost function was written with autograd in mind which I think is already quite a strong assumption, and, if violated, results in an obscure error message if we try to differentiate such a function in the autograd backend. While we could simply iterate over all possible backends for callables and try-except until we find one which succeeds, this approach would mask possible issues with cost functions. For example, say a user tries to multiply two arrays of incompatible dimensions, this approach would result in an error message claiming we couldn't find a suitable autodiff backend while masking the bigger underlying issue of the erroneous cost function. Instead, I think we should lower the abstraction level regarding the way cost functions (as well as gradients and Hessians) are defined in pymanopt. The idea is essentially to require users to be a little more explicit with what backend they want to use. Before outlining my proposed changes, let me first describe the other two (admittedly small) issues I noticed.

1. If provided, we currently assume that (e)grad/e(hess) are callables. This prevents users from defining derivatives the same way they might define cost functions, i.e., as symbolic theano/tensorflow expressions. While this might not be the biggest issue (at least I can't think of an obvious situation where this might be desirable), I think it makes for a somewhat inconsistent or at least unintuitive API. Note that this issue in and of itself is easily fixed by adding the necessary checks to the egrad/ehess methods in the `Problem` class, and simply compiling such objects with the appropriate backend if necessary.
2. Again, while I'm not quite sure of the particular use case for this, the current implementation does not allow to mix different types of cost functions and derivatives. For instance, it's not possible to define the cost function as theano graph, and the gradient as tensorflow graph. This is due to the fact that we only accept a single `arg` keyword argument in the `Problem` class, meaning that our backend detection logic will always try to match a symbolic expression with an incompatible argument for differentiation if cost and derivatives use different frameworks. The only clean way to decouple this dependency is to somehow equip the computation graphs with the necessary arguments for compilation and differentation. Adding something like `egrad_arg`, etc. to the `Problem` constructor would just add more complexity to the already complex argument list of the `Problem` class.

## Option 1: decorators for callables

My initial idea to circumvent the pytorch/autograd callable issue was to introduce decorators for cost functions which aren't defined as symbolic expressions. These decorators would effectively just add a `.backend` attribute to the decorated function as a hint for our autodiff logic what type of function we're dealing with. This would then look something like this:

```python
from autograd import numpy as np
import torch

from pymanopt import decorators

@decorators.autograd
def cost(x):
    return np.inner(x, np.dot(A, x))

@decorators.pytorch
def egrad(x):
    return 2 * torch.matmul(torch.from_numpy(A), x)
```

The good thing about this solution is that it requires only mild user-facing API changes in that autograd functions now need to be annotated with the provided decorator. Note that this approach won't allow to fix issue 2. I outlined above.

The other option would be more invasive.

## Option 2: function wrapper classes

The basic idea here is to introduce a base class `Function` to wrap cost functions, gradients and Hessians in. For each backend, we'd then implement a separate subclass which essentially replaces the backend logic. In practice, I envision something like

```python
from autograd import numpy as np
import theano.tensor as T

from pymanopt import AutogradFunction, TheanoFunction

@AutogradFunction
def cost(x):
  return np.inner(x, np.dot(A, x))

# or alternatively
cost = AutogradFunction(lambda x: np.inner(x, np.dot(A, x)))

x = T.vector()
egrad = TheanoFunction(2 * T.dot(A, x), x)
```

The way this would work is that each `Function` subclass implements a constructor accepting a callable or symbolic expression, and an optional `arg` parameter, as well as a `__call__` method to turn it into a callable. This way, classes which only require a callable (such as pytorch and autograd functions) can use the respective class as a simple decorator, while symbolic expression frameworks simply wrap the function expressions in the respective class.

I'm currently in the middle of preparing a PR for the first option to demonstrate how this implementation would work. This would also lay the groundwork for option 2 should we decide to go that route. Of course, if you guys think it's not worthwhile to support pytorch to begin with and don't feel comfortable with such invasive API changes, let me know.

**Comment by nkoep** [comment]
Closing this after the merge of https://github.com/pymanopt/pymanopt/pull/67.
